### PR TITLE
Parallel

### PR DESCRIPTION
**Pull request by jialinding**
Parallel merge bench and RandomSketch

### CODE REVIEW

**Review by edgan8 — CHANGES_REQUESTED** [review]

File: javamsketch/quantilebench/src/main/java/yahoo/QuantilesHelper.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
+package yahoo;
+
+/**
+ * Common static methods for quantiles sketches
+ */
+public class QuantilesHelper {
```
[the remark this answers is not in the document]
Comment (jialinding) [reply javamsketch/quantilebench/src/main/java/yahoo/QuantilesHelper.java]: I only copied files when they were needed to get rid of an error.

File: javamsketch/quantilebench/src/main/java/sketches/TDigestSketch.java [hunk 2]
```diff
@@ -59,15 +65,56 @@ public void add(double[] data) {
...
-            TDigestSketch ts = (TDigestSketch)s;
+        for (int i = startIndex; i < endIndex; i++) {
+            TDigestSketch ts = (TDigestSketch) sketches.get(i);
             newTD.add(ts.td);
         }
         return this;
     }
 
+    @Override
+    public QuantileSketch parallelMerge(ArrayList<QuantileSketch> sketches, int numThreads) {
```
[the remark this answers is not in the document]
Comment (jialinding) [reply javamsketch/quantilebench/src/main/java/sketches/TDigestSketch.java]: Yee

File: javamsketch/quantilebench/src/main/java/sketches/RandomSketch.java [hunk 3]
```diff
@@ -0,0 +1,558 @@
...
+            }
+
+            curBuffer.add(x);
+            if (activeLevel == 0) {
+                totalWeight++;
+            } else {
+                totalWeight += sampleBlockLength;
+            }
+            if (curBuffer.size() == s) {
+                Collections.sort(curBuffer);
```
[the remark this answers is not in the document]
Comment (jialinding) [reply javamsketch/quantilebench/src/main/java/sketches/RandomSketch.java]: Yeah that makes it more efficient when combining buffers. This is the way it's described in the paper and in the coolwanglu code.

File: javamsketch/quantilebench/src/main/java/sampling/SamplingUtil.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+import static com.yahoo.sketches.BoundsOnBinomialProportions.approximateUpperBoundOnP;
+
+import java.util.Random;
+
+/**
+ * Common utility functions for the sampling family of sketches.
+ *
+ * @author Jon Malkin
+ */
+final class SamplingUtil {
```
[the remark this answers is not in the document]
Comment (jialinding) [reply javamsketch/quantilebench/src/main/java/sampling/SamplingUtil.java]: good point

[the remark this answers is not in the document]
Comment (jialinding) [reply javamsketch/quantilebench/src/main/java/sampling/SamplingUtil.java]: actually some functions still use SamplingUtil for non-sampling purposes

File: javamsketch/quantilebench/src/main/java/sampling/SamplingUtil.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+    /**
+     * Number of standard deviations to use for subset sum error bounds
+     */
+    private static final double DEFAULT_KAPPA = 2.0;
+
+    public static final Random rand = new Random();
```
[the remark this answers is not in the document]
Comment (jialinding) [reply javamsketch/quantilebench/src/main/java/sampling/SamplingUtil.java]: ok

**Review by edgan8 — APPROVED** [review]

**Pull request merged**